crates/deltalake-core/src/delta_datafusion/cdf/scan.rs

```diff
@@ -0,0 +1,318 @@
+                        commit_info.replace(ci);
+                    }
+                    _ => {}
+                }
+            }
+
+            if !cdc_actions.is_empty() {
+                change_files.push(CdcDataSpec {
+                    version,
+                    timestamp: ts.unwrap_or(0),
+                    actions: cdc_actions,
+                    commit_info: commit_info.cloned(),
+                })
+            } else {
+                let add_actions = version_actions
+                    .iter()
+                    .filter_map(|a| match a {
+                        Action::Add(a) if a.data_change => Some(a.clone()),
+                        _ => None,
+                    })
+                    .collect::<Vec<Add>>();
+
+                let remove_actions = version_actions
+                    .iter()
+                    .filter_map(|a| match a {
+                        Action::Remove(r) if r.data_change => Some(r.clone()),
+                        _ => None,
+                    })
+                    .collect::<Vec<Remove>>();
+
+                if !add_actions.is_empty() {
+                    add_files.push(CdcDataSpec {
+                        version,
+                        timestamp: ts.unwrap_or(0),
+                        actions: add_actions,
+                        commit_info: commit_info.cloned(),
+                    });
+                }
+
+                if !remove_actions.is_empty() {
+                    remove_files.push(CdcDataSpec {
+                        version,
+                        timestamp: ts.unwrap_or(0),
+                        actions: remove_actions,
+                        commit_info: commit_info.cloned(),
+                    });
+                }
+            }
+        }
+
+        Ok((change_files, add_files, remove_files))
+    }
+
+    fn get_add_action_type() -> Option<ScalarValue> {
+        Some(ScalarValue::Utf8(Some(String::from("insert"))))
+    }
+
+    pub async fn scan(&self) -> DeltaResult<SendableRecordBatchStream> {
+        let (cdc, add, _remove) = self.determine_files_to_read().await?;
+        let ctx = SessionContext::new();
+        register_store(self.log_store.clone(), ctx.state().runtime_env().clone());
+
+        let schema_fields: Vec<Field> = self
+            .schema
+            .all_fields()
+            .into_iter()
+            .filter(|f| !self.partition_values.contains(f.name()))
+            .cloned()
+            .collect();
+
+        let this_partition_values = self
+            .partition_values
+            .iter()
+            .map(|name| self.schema.field_with_name(name).map(|f| f.to_owned()))
+            .collect::<Result<Vec<_>, ArrowError>>()?;
+
+        // Setup for the Read Schemas of each kind of file, CDC files include commit action type so they need a slightly
+        // different schema than standard add file reads
+        let cdc_file_schema = create_cdc_schema(schema_fields.clone(), true);
```

Comment:
Here you add CHANGE_TYPE_COL but in CDC_PARTITION_SCHEMA it's not there? I am a bit confused about this

Reply:
because in cdc partitions those files already have the change type where as adds don't have them so the schema needs to account for this.